src/main/java/com/google/devtools/build/lib/worker/WorkerProxy.java

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.sandbox.SandboxHelpers;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A proxy that talks to the multiplexers
+ */
+final class WorkerProxy extends Worker {
+  private Integer workerHash;
+  private ByteArrayOutputStream request;
+  private WorkerMultiplexer workerMultiplexer;
+  private Thread shutdownHook;
+
+  WorkerProxy(WorkerKey workerKey, int workerId, Path workDir, Path logFile) {
+    super(workerKey, workerId, workDir, logFile);
+    this.workerHash = workerKey.hashCode();
+    this.request = new ByteArrayOutputStream();
+    try {
+      this.workerMultiplexer = WorkerMultiplexerManager.getInstance(workerHash);
+    } catch (InterruptedException e) {
+      // We can't do anything here.
```

Comment:
Move this to a constructor argument and grab the instance when `WorkerProxy` is constructed, then you don't need this try/catch.

Reply:
Good suggestion. Is it better to throw exception in `WorkerFactory`?

https://github.com/lucidsoftware/bazel/blob/45854c3efbb18a0d2864e4fa9ebbfdde9c57b77b/src/main/java/com/google/devtools/build/lib/worker/WorkerFactory.java#L55